feat: AEP 7: Queue system interface design (#51)
aneoconsulting/ArmoniK.Community: AEP/aep-00007.md

It the task completes successfully, the message is acknowledged after tasks results are uploaded to the storage and tasks created by the task in processing are submitted.
If the queue service loses connection with the processing agent, the message is redelivered, ensuring task execution even in the presence of errors.

Additionally, tasks that are pending can be released when a long-running task is in progress.
ArmoniK initiates the processing of a new message only when the current task has been dispatched to a worker, starting the acquisition of a new task during the processing of the previous one.
This design allows to have a task ready for execution as soon as the previous task ends.
Tasks undergoing retry are treated as entirely new tasks, which simplifies tracking and execution, generating new messages accordingly.

Message uniqueness is not required as it is managed elsewhere.
Furthermore, message scheduling must be handled within the queue service to accommodate prioritization mechanisms.
The system should also allow for the seamless integration of new plugins to enhance flexibility and adaptability.

## Possible Approaches for the Interface

> Several approaches can be considered for implementing the message processing interface, including event-driven and pull-based mechanisms. Given the need to control pipelining—managing the number of concurrently processed messages and determining when processing begins—both approaches are functionally equivalent. In practice, polling would be used on an internal queue to regulate message processing.

Reviewer: ```suggestion
Several approaches can be considered to implement the message processing interface, including event-driven and pull-based mechanisms. Given the need to control pipelinin, managing the number of concurrently processed messages and determining when processing begins, are functionally equivalent. In practice, polling would be used on an internal queue to regulate message processing.
```

Below is an overview of how major market solutions implement message retrieval mechanisms:

> | Solution      | Approche |

Reviewer 2: Approach

| ------------- | -------- |
| AWS SQS       | pull     |
| ActiveMQ      | both     |
| RabbitMQ      | both     |
| Google PubSub | both     |
| Pulsar        | both     |

## Trade-offs and Constraints

> Rewriting the codebase to adopt a fully event-driven, push-based pattern is not feasible given the complexity and scale of such a transformation. The current design prioritizes simplicity and robustness over a pure event-driven model. Considering the long duration of tasks and the cost of orchestration, some inefficiencies during message reception are acceptable. These inefficiencies are preferable to the development overhead and technical debt involved in transforming the existing pull-based mechanism.

Reviewer: ```suggestion
The current design prioritizes simplicity and robustness over a pure event-driven model. Considering the long duration of tasks and the cost of orchestration, some inefficiencies during message reception are acceptable. These inefficiencies are preferable to the development overhead and technical debt involved in transforming the existing pull-based mechanism. Rewriting the codebase to adopt a fully event-driven, push-based pattern would implied to rewrite most of the orchestrator.
```

> The technical limitations of the underlying queueing service also play a role in the design decisions. SQS, for instance, does not provide an API for push-based message reception, which necessitates the use of a pull-based approach. Additionally, past attempts to use RabbitMQ for push-based message handling revealed significant stability issues, such as connection losses and inconsistent message processing. These challenges have further reinforced the decision to rely on the pull-based model, which offers greater reliability and predictability.

Reviewer: ```suggestion
The technical limitations of the underlying queueing service also play an important role in the design decisions. SQS, for instance, does not provide an API for push-based message reception, which necessitates to use of a pull-based approach. Additionally, past attempts to use RabbitMQ for push-based message handling revealed significant stability issues, such as connection losses and inconsistent message processing. These challenges have further reinforced the decision to rely on the pull-based model, which offers greater reliability and predictability.
```

> We can consider converting a push-based reception to messages into our pull-based interfaces. However, this implementation requires to be as stable as the current one. Pull requests are welcome to improve the implementation of our queue system adaptors.

Reviewer: ```suggestion
We can consider converting a push-based reception to messages into our pull-based interfaces. Naturally, this implementation requires to be proved as stable and as scalable than the current one. Pull requests are welcome to improve the implementation of our queue system adaptors.
```

## Current Interface

Queue interfaces can be found in our dotnet package [ArmoniK.Core](https://www.nuget.org/packages/ArmoniK.Core.Base).

> It contains everything needed to implement queue plugins that can be dynamically loaded by ArmoniK.Core allowing users to implement plugins that match their requirements.

Reviewer 2: I'd write "The package provides the elements necessary to implement queue plugins ... "

> An agent from a partition calls the `PullMessagesAsync` to get messages representing tasks from the associated partition.

Reviewer 2: ... the `PullMessagesAsync`  method

The partition is given outside the `IPullQueueStorage` and should be passed to the implementation through other means.

> It is usually done with .Net options system and environment variables.

Reviewer: ```suggestion
It is usually done with `.Net` options system and environment variables.
```

> This also means there is no uniformization to set up the partition and it will depend on the implementation of the interface.

Reviewer: will that stay like that? And it is something that we want?

Author: No. No

```csharp
/// <summary>
///   Interface to retrieve messages from the queue
/// </summary>
public interface IPullQueueStorage : IQueueStorage
{
  /// <summary>
  ///   Gets messages from the queue
  /// </summary>
  /// <param name="nbMessages">Number of messages to retrieve</param>
  /// <param name="cancellationToken">Token used to cancel the execution of the method</param>
  /// <returns>
  ///   Enumerator allowing async iteration over the message queue
  /// </returns>
  IAsyncEnumerable<IQueueMessageHandler> PullMessagesAsync(int               nbMessages,
                                                           CancellationToken cancellationToken = default);
}
```

> The `PullMessagesAsync` method returns an `IAsyncEnumerable<IQueueMessageHandler>` where the `IQueueMessageHandler` is the following interface.

Reviewer 2: I'd write: "where the `IQueueMessageHandler` is an interface representing the life cycle of a message:" 

and move 

The `QueueMessageStatus` is set by ArmoniK during the execution of the task.
Then, the `DisposeAsync` method inherited from `IAsyncDisposable` is used to process the message from the queue by acknowledging it, not acknowledging it, or requeuing it depending on the status.

after the interface definition.

  /// </summary>
  QueueMessageStatus Status { get; set; }

  /// <summary>
  ///   Date of reception of the message
  /// </summary>
  DateTime ReceptionDateTime { get; init; }
}
```

```csharp
/// <summary>
///   Represents the status of a queue message
/// </summary>
public enum QueueMessageStatus
{
  /// <summary>
  ///   Message is waiting for being processed.
  /// </summary>
  Waiting,

  /// <summary>
  ///   Message processing has failed. The message should be put back at the begin of the queue.
  /// </summary>
  Failed,

  /// <summary>
  ///   The message is being processed.
  /// </summary>
  Running,

  /// <summary>
  ///   Task is not ready to be processed. The message should be put at the end of the queue.
  /// </summary>
  Postponed,

  /// <summary>
  ///   The message has been processed. It can safely be removed from the queue.
  /// </summary>
  Processed,

  /// <summary>
  ///   The message processing has been cancelled. the message can safely be removed from the queue.
  /// </summary>
  Cancelled,

  /// <summary>
  ///   Message has been retried too many times and is considered as poisonous for the queue
  /// </summary>
  Poisonous,
}
```

> Push interface is a lot simpler.

Reviewer 2: The Push interface

> Push interface is a lot simpler.

Reviewer: When I am reading it , I am a little bit confused since the previous part where saying that we do not a push in armonik. Perhaps it is clear for people knowing the in-out of armonik but it is confusing for someone else I think. Perhaps explicitely saying which process of the orchestrator is using it will help the comprehension (subtasking and resubmitt)

Author: There are 2 kinds of push:
- push messages from the queue to ArmoniK -> debate is up for this one.
- push messages from ArmoniK to the queue -> we cannot avoid this one.

  /// <returns>
  ///   Task representing the asynchronous execution of the method
  /// </returns>
  public Task PushMessagesAsync(IEnumerable<MessageData> messages,
                                string                   partitionId,
                                CancellationToken        cancellationToken = default);
}
```

```csharp
/// <summary>
///   Data structure to hold message data
/// </summary>
/// <param name="TaskId">Unique identifier of the task</param>
/// <param name="SessionId">Unique name of the session to which this message belongs</param>
/// <param name="Options">Task options</param>
public record MessageData(string      TaskId,
                          string      SessionId,
                          TaskOptions Options);
```

## Issues with the current interface

There are a few issues from the current interfaces:
- The split of the interface into two due to the configuration. The partition should be given in the pull method directly and uniformize partition selection.
- Clarify message processing instead of relying on the `DisposeAsync` method from the `IQueueMessageHandler`.

# Conclusion

> The architecture reflects a deliberate balance between design complexity, operational requirements, and system constraints. By combining a historical interface with strict control over message processing and lifecycle management, ArmoniK achieves a reliable and scalable system. Although the system does not adopt a fully event-driven approach, it remains robust and resilient, meeting the demands of long-running tasks and orchestration challenges. The trade-offs in this design are justified by the system’s operational stability and its ability to handle large-scale workloads efficiently.

Reviewer: _it remains robust and resilient,_

Some people are arguing that it is not the case and even more that most of the problems encounter with armonik are because of this choice. Not sure that we should leave this sentence as it is.

Author: I think it is not a problem with the push/pull approach method to receive messages, but an issue with messages insertion into ActiveMQ. It is either a problem with ActiveMQ or implementation of the AMQP protocol. Otherwise, we would have a lot of similar issues with SQS and PubSub, which is not the case.

Message uniqueness is not required as it is managed elsewhere.
Furthermore, message scheduling must be handled within the queue service to accommodate prioritization mechanisms.
The system should also allow for the seamless integration of new plugins to enhance flexibility and adaptability.

## Possible Approaches for the Interface

Several approaches can be considered to implement the message processing interface, including event-driven and pull-based mechanisms. Given the need to control pipelining in order to manage the number of concurrently processed messages and determining when processing begins, both approaches are functionally equivalent. In practice, polling would be used on an internal queue to regulate message processing.

Below is an overview of how major market solutions implement message retrieval mechanisms:

| Solution      | Approach |
| ------------- | -------- |
| AWS SQS       | pull     |
| ActiveMQ      | both     |
| RabbitMQ      | both     |
| Google PubSub | both     |
| Pulsar        | both     |

## Trade-offs and Constraints

>  The current design prioritizes simplicity and robustness over a pure event-driven model. Considering the long duration of tasks and the cost of orchestration, some inefficiencies during message reception are acceptable. These inefficiencies are preferable to the development overhead and technical debt involved in transforming the existing pull-based mechanism. Rewriting the codebase to adopt a fully event-driven, push-based pattern is not feasible given the complexity and scale of such a transformation.
> 
> The technical limitations of the underlying queueing service also play an important role in the design decisions. SQS, for instance, does not provide an API for push-based message reception, which necessitates to use of a pull-based approach. Additionally, past attempts to use RabbitMQ for push-based message handling revealed significant stability issues, such as connection losses and inconsistent message processing. These challenges have further reinforced the decision to rely on the pull-based model, which offers greater reliability and predictability.
> 
> We can consider converting a push-based reception to messages into our pull-based interfaces. Naturally, this implementation requires to be proved as stable and as scalable as the current one. Pull requests are welcome to improve the implementation of our queue system adaptors.

Reviewer 3: ```suggestion
### Event-driven approach (push)
Using an event-driven approach allows the Queue System to push the messages to the Scheduling Agent. The control of "when the messages arrives in the Scheduling Agent" is moved to the Queue system. This allows to reduce the latency between the submission of a message in the Queue System and its dispatch in the Scheduling Agent. On the other hand, managing the number of messages received by the Scheduling Agent is more complex, or impossible if the number of messages allowed in each Scheduling Agent can change frequently over time. 

### Pull based approach
Using a pull-based approach, the Scheduling Agent has to call the Queue System to try and get a new message. This allows the Scheduling Agent to control when a new message is obtained. However, this increases the latency to get new messages.

Note that in the Scheduling Agent pipelined implementation, polling is used on internal queues to regulate the number of message being in each step of processing.

### Choice
The technical limitations of the underlying queueing service also play an important role in the design decisions. SQS, for instance, does not provide an API for push-based message reception, which necessitates to use of a pull-based approach. A library can expose a push-based interface while using a pull-based Queue System (or vice-versa).

In the end, we believe that the choice made should be the result of two main constraints : 
* a pipelined implementation 
* a need to minimize the number of messages stuck in the steps prior the task execution

These constraints are more easily taken into account with a pull-based approach as the information to decide whether or not a new message should enter the pipeline are available in the Scheduling Agent rather than the Queue system.

Additionally, past experiments to use RabbitMQ for push-based message handling revealed significant stability issues, such as connection losses and inconsistent message processing. These challenges have further reinforced the decision to rely on the pull-based model, which offers greater reliability and predictability.
```

  /// </summary>
  [Obsolete("ArmoniK now manages loss of link with the queue")]
  CancellationToken CancellationToken { get; set; }

  /// <summary>
  ///   Id of the message
  /// </summary>
  string MessageId { get; }

  /// <summary>
  ///   Task Id contained in the message
  /// </summary>
  string TaskId { get; }

  /// <summary>
  ///   Status of the message. Used when the handler is disposed to notify the queue.
  /// </summary>
  QueueMessageStatus Status { get; set; }

  /// <summary>
  ///   Date of reception of the message
  /// </summary>
  DateTime ReceptionDateTime { get; init; }
}
```

The `QueueMessageStatus` is set by ArmoniK during the execution of the task.
Then, the `DisposeAsync` method inherited from `IAsyncDisposable` is used to process the message from the queue by acknowledging it, not acknowledging it, or requeuing it depending on the status.
The statuses are defined as follow:

```csharp
/// <summary>
///   Represents the status of a queue message
/// </summary>
public enum QueueMessageStatus
{
  /// <summary>
  ///   Message is waiting for being processed.
  /// </summary>
  Waiting,

  /// <summary>

>   ///   Message processing has failed. The message should be put back at the begin of the queue.

Reviewer 4: I am not sure about this one.

  ///   Task representing the asynchronous execution of the method
  /// </returns>
  public Task PushMessagesAsync(IEnumerable<MessageData> messages,
                                string                   partitionId,
                                CancellationToken        cancellationToken = default);
}
```

```csharp
/// <summary>
///   Data structure to hold message data
/// </summary>
/// <param name="TaskId">Unique identifier of the task</param>
/// <param name="SessionId">Unique name of the session to which this message belongs</param>
/// <param name="Options">Task options</param>
public record MessageData(string      TaskId,
                          string      SessionId,
                          TaskOptions Options);
```

## Issues with the current interface

There are a few issues from the current interfaces:
- The split of the interface into two due to the configuration. The partition should be given in the pull method directly and uniformize partition selection.
- Clarify message processing instead of relying on the `DisposeAsync` method from the `IQueueMessageHandler`.

These issues will be addressed in a new version of these interfaces.

# Conclusion

> The architecture reflects a deliberate balance between design complexity, operational requirements, and system constraints. By combining a historical interface with strict control over message processing and lifecycle management, ArmoniK achieves a reliable and scalable system. Although the system does not adopt a fully event-driven approach, it remains robust and resilient, meeting the demands of long-running tasks and orchestration challenges. The trade-offs in this design are justified by the system’s operational stability and its ability to handle large-scale workloads efficiently.

Reviewer 3: ```suggestion
The architecture reflects a deliberate balance between design complexity, operational requirements, and system constraints. The interface reflects an architecture that allows a strict and precise control over message processing and lifecycle management. This allows ArmoniK to achieve a reliable and scalable system. The trade-offs in this design are justified by the system’s operational stability and its ability to handle large-scale workloads efficiently.
```

```Although the system does not adopt a fully event-driven approach, it remains robust and resilient, meeting the demands of long-running tasks and orchestration challenges.``` has been removed as it could be interpreted that a full event-driven approach should be a target. This is not the case.

# AEP 7: Queue system interface design

|                     |ArmoniK Enhancement Proposal|
|---:                 |:--- |
| **AEP**             | 7 |
| **Title**           | Queue system interface design |
| **Author**          | Jérôme Gurhem <<jgurhem@aneo.fr>> |
| **Status**          | Active |
| **Type**            | Standard |
| **Creation Date**   | 2025-01-15 |

# Abstract

This proposal outlines the rationale and design decisions behind the current message processing and lifecycle management system in ArmoniK. It highlights the primary reasons for adopting a pull-based model, discusses design constraints, and evaluates the trade-offs involved in ensuring reliability and robustness for long-running tasks.

# Motivation

The main motivation for the current architecture is to maintain a historical interface that provides comprehensive traceability and monitoring of tasks and message processing. This interface is central to operational observability and is a key requirement for the system.

# Rationale

## Strict Control of Message Processing

> The system is designed to enforce strict control over the number of messages being processed and their lifecycle. Each message in the system represents an executable task with a well-defined processing flow. An executable task is a task which dependencies were completed, thus being queued for execution. When a message is received, the corresponding task begins execution. The message is acknowledged only after the task completes, regardless of whether it succeeds or fails. In cases where the queueing service loses its connection with the agent before the message is acknowledged, the system ensures reliability by redelivering the message. This guarantees that tasks are not lost and are executed even when errors or interruptions occur.

Reviewer 2: "... is a task which dependencies were completed" => replace which by whose

> Retries for tasks are handled by treating them as new tasks. Each retry is introduced as a new message, which simplifies tracking and execution. Messages are acknowledged when the task completes successfully and subtasks are submitted. ArmoniK processes new messages only after the current task has been dispatched to a worker, starting the acquisition of a new task during the processing of the previous one. This design allows to have a task ready for execution as soon as the previous task ends.

Reviewer 2: I'd rewrite  "Messages for retry tasks are acknowledged when ..." just to be more explicit that in this case we do care about task completion success.  And the reference to subtasking confuses me, since is not at all mentioned on the previous paragraph addressing "normal tasks".
